### ISSUE TITLE
Package request: o3de

### ISSUE THREAD

**Issue by Ashvith10** [NixOS/nixpkgs#255307]
**Project description**

 Open 3D Engine (O3DE) is an Apache 2.0-licensed multi-platform 3D engine that enables developers and content creators to build AAA games, cinema-quality 3D worlds, and high-fidelity simulations without any fees or commercial obligations. 

**Metadata**

* homepage URL: https://o3de.org/
* source URL: https://github.com/o3de/o3de
* license: mit, apache-2.0, unfree
* platforms: unix

**Comment by Ashvith10** [comment]
I can try working on a derivation, but I'm not sure if I will be able to build this locally.

**Comment by QushyQushy** [comment]
Any news on this?

**Comment by Ashvith10** [comment]
@QushyQushy there's actually a bunch of issues with building o3de, and most of it has to do with how the bash scripts fail, because of the shebang in Python and Shell script respectively. I don't have a lot of experience with using patches in Nix expressions. This project tries to download it's own Python binaries, which isn't reproducible, I think. Also, my system isn't powerful enough to build this. It would be better if someone with proper Nix proficiency looks into this. However, I can give this a last try.

**Comment by QushyQushy** [comment]
> @QushyQushy there's actually a bunch of issues with building o3de, and most of it has to do with how the bash scripts fail, because of the shebang in Python and Shell script respectively. I don't have a lot of experience with using patches in Nix expressions. This project tries to download it's own Python binaries, which isn't reproducible, I think. Also, my system isn't powerful enough to build this. It would be better if someone with proper Nix proficiency looks into this. However, I can give this a last try.
> 
> [Here's](https://www.docs.o3de.org/docs/welcome-guide/setup/setup-from-github/building-linux/) the build instructions for Linux, if you want to try them for yourself.

Thank you so much! I have zero familiarity with the Nix build system unfortunately. I thought it would be as simple as downloading the .deb file and extracting it in the proper places but I have no idea in reality.

**Comment by polypoyo** [comment]
Can we reopen this? I don't want to make a duplicate issue

**Comment by Ashvith10** [comment]
@polypoyo I've re-opened this.

**Comment by tanshihaj** [comment]
I've tried to build o3de using nix + nixpkgs and here some thoughts about it. Even though build process pretty good documented o3de has one big issue regarding to building process: [O3DE Packages](https://docs.o3de.org/docs/user-guide/build/packages). As well as other big c++/cmake projects this project invented its own cross-platform dependency management tooling. In general terms CMake scripts (engine itself, gems and project files) contains O3DE Packages declarations (here is the example from engine scripts: https://github.com/o3de/o3de/blob/c753ddf5c25c8cc7aa44845afd035e3996962e81/cmake/3rdParty/Platform/Linux/BuiltInPackages_linux_x86_64.cmake#L20-L48) with tar hashes. CMake functions tries to download, check hash, unpack and check unpacked file hashes of these O3DE Packages into `$LY_3RDPARTY_PATH` folder. CMake targets then can refer to these O3DE packages as `3rdParty::ZLIB`, for example.

This approach obviously conflicts with nix + nixpkgs because:
1. nix build environment with enabled sandboxing will not allow build script to access network.
2. Even if we will pre-download all these dependencies binaries from these packages will not run in nixos due to incorrect interpreter path and incorrect RPATH.
3. Even if we will pre-download all required dependencies and run patchelf on them for engine code what to do with O3DE packages declared in gems and projects?

I see several solutions to fix this issue, one worse than other:
1. Pre-download and pre-patch all required dependencies for engine itself, require special environment for packages with all required dependencies pre-downloaded and pre-patched. Kinda true nix-way but it is pretty hard-to-use for package users.
2. Pre-download binaries for engine and hack somehow cmake functions in https://github.com/o3de/o3de/blob/development/cmake/3rdParty.cmake and https://github.com/o3de/o3de/blob/development/cmake/3rdPartyPackages.cmake so they will automatically patch binaries after downloading.

Maybe somebody here has more experience with dealing cmake-based package managers?

NB: it is just build-related issues, I think there can be bunch of other questions regarding to the usability: engine registration, packaging project for release, etc. I think this questions should be discussed separately.

**Comment by nick-l-o3de** [comment]
I hope someone can pick this up.  The "source" for all these 3p packages are hosted in https://github.com/o3de/3p-package-source/ and you'll  find that most of them consist of just a standard build script that downloads the original source code from wherever it lives, and patches it with a patch thats also present there, then builds it with specific flags.  Most builds are completely normal.  The only really complicated ones are python and Qt.

On the other hand, the package system on the O3DE side is fairly lenient - it only goes to download packages if the target is missing.  So for example, if the OS provides one of those packages, a nix-specific CMAke script could go import those package targets before invoking the rest of the O3DE build scripts, and would use those instead.

In other words, 3rdPartyPackages.cmake will only attempt to download packages if they are not defined at the time someone actually declares a dependency on them.  If `3rdParty::Foo` already exists (becuase a CMake script `aliased` it with `Foo` from the system), it will use that instead.

**Comment by Ulrick28** [comment]
What do we gain by putting resources into supporting Nix? Apologies for my naivety here.

**Comment by Ashvith10** [comment]
@Ulrick28 I don't use NixOS anymore, and I am also not an expert in this topic - so apologies if my points are not that strong or convincing. I use Guix, which is related to Nix, but has diverged in some aspect. In Guix, we have the option to pick a commit  outside of the pulled channel, something like:

```console
# This snippet and commit is made-up
$ guix time-machine \
 --branch=v1.3.0 \
 --commit=ae54bg \
 -- shell o3de-engine \
 --pure
```
So, in a way, I can have access to multiple version of a package expression, that may have pointed to an older version of the same application in an purely isolated environment. I could also run this in a container using:

```console
# Also a made-up snippet
$ guix time-machine \
 --commit=be5yt2 \
 -- shell o3de-engine \
 -CPW
```

If I wanted to, I can also point to my personal Guix channel:

```console
$ guix time-machine \
 --url=$HOME/Desktop/o3de-engine \ # or https://github.com/some-user/o3de-engine
 -- shell o3de-engine
```

Benefits include:
- ephemeral shells with varying levels of isolation (pure, container, vm)
- time-travel to older, newer or forked versions
- caching of builds in the store with it's own unique hash
- no residual files left behind in the `/bin`, `/usr/bin` or `/etc` file
- reduced re-build time (with modularization), as the other components are cached in store
- custom package build can be deal with package transformations

Nix also has it's own command-equivalent to what I have mentioned, which I cannot recall right now.
